### PR TITLE
`ruff server`: Ruff configuration from client settings overrides project configuration

### PR DESCRIPTION
**Pull request by snowsignal**
## Summary

This is a follow-up to https://github.com/astral-sh/ruff/pull/10984 that implements configuration resolution for editor configuration. By 'editor configuration', I'm referring to the client settings that correspond to Ruff configuration/options, like `preview`, `select`, and so on. These will be combined with 'project configuration' (configuration taken from project files such as `pyproject.toml`) to generate the final linter and formatter settings used by `RuffSettings`. Editor configuration takes priority over project configuration.

In a follow-up pull request, I'll implement a new client setting that allows project configuration to override editor configuration, as per [this issue](https://github.com/astral-sh/ruff-vscode/issues/425).

## Review guide

The first commit, e38966d8843becc7234fa7d46009c16af4ba41e9, is just doing re-arrangement so that we can pass the right things to `RuffSettings::resolve`. The actual resolution logic is in the second commit, 0eec9ee75c10e5ec423bd9f5ce1764f4d7a5ad86. It might help to look at these comments individually since the diff is rather messy.

## Test Plan

For the settings to show up in VS Code, you'll need to checkout this branch: https://github.com/astral-sh/ruff-vscode/pull/456.

To test that the resolution for a specific setting works as expected, run through the following scenarios, setting it in project and editor configuration as needed:

| Set in project configuration? | Set in editor configuration?                     | Expected Outcome                                                                         |
|-------------------------------|--------------------------------------------------|------------------------------------------------------------------------------------------|
| No                            | No                                               | The editor should behave as if the setting was set to its default value.                  |
| Yes                           | No                                               | The editor should behave as if the setting was set to the value in project configuration. |
| No                            | Yes                                              | The editor should behave as if the setting was set to the value in editor configuration.  |
| Yes                           | Yes (but distinctive from project configuration) | The editor should behave as if the setting was set to the value in editor configuration.  |

An exception to this is `extendSelect`, which does not have an analog in TOML configuration. Instead, you should verify that `extendSelect` amends the `select` setting. If `select` is set in both editor and project configuration, `extendSelect` will only append to the `select` value in editor configuration, so make sure to un-set it there if you're testing `extendSelect` with `select` in project configuration.

### CODE REVIEW

**Comment by snowsignal** [comment]
I'm working on adding some snapshot tests, but the PR should be ready for review regardless.

File: crates/ruff_server/src/session/workspace/ruff_settings.rs [hunk 1]
```diff
@@ -47,20 +71,18 @@ impl RuffSettingsIndex {
             .map(DirEntry::into_path)
         {
             if let Some(pyproject) = settings_toml(&directory).ok().flatten() {
-                let Ok(settings) = ruff_workspace::resolver::resolve_root_settings(
+                let Ok(configuration) = ruff_workspace::resolver::resolve_configuration(
```
[the remark this answers is not in the document]
Comment (snowsignal) [reply crates/ruff_server/src/session/workspace/ruff_settings.rs:74]: Oh, good catch!

**Review by MichaReiser — APPROVED** [review]
Thank you

**Review by dhruvmanila — APPROVED** [review]

**Comment by snowsignal** [comment]
I'll implement snapshot tests [as a follow-up](https://github.com/astral-sh/ruff/issues/11112)

**Pull request merged**